src/analysis/processing/qgsalgorithmfillsinkswangliu.cpp

```diff
@@ -0,0 +1,438 @@
+          steepestDirection = direction;
+        }
+      }
+    }
+  }
+
+  return steepestDirection;
+}
+
+struct CFillSinks_WL_Node
+{
+    int row = 0;
+    int col = 0;
+    double spill = 0;
+};
+
+class CompareGreater
+{
+  public:
+    bool operator()( CFillSinks_WL_Node n1, CFillSinks_WL_Node n2 ) const
+    {
+      return n1.spill > n2.spill;
+    }
+};
+
+typedef std::vector< CFillSinks_WL_Node > nodeVector;
+typedef std::priority_queue< CFillSinks_WL_Node, nodeVector, CompareGreater > PriorityQ;
+
+QVariantMap QgsFillSinksWangLiuAlgorithm::processAlgorithm( const QVariantMap &parameters, QgsProcessingContext &context, QgsProcessingFeedback *feedback )
+{
+  const QString createOptions = parameterAsString( parameters, QStringLiteral( "CREATE_OPTIONS" ), context ).trimmed();
+
+  const QString filledDemOutputFile = parameterAsOutputLayer( parameters, QStringLiteral( "OUTPUT_FILLED_DEM" ), context );
+  const QString flowDirectionsOutputFile = parameterAsOutputLayer( parameters, QStringLiteral( "OUTPUT_FLOW_DIRECTIONS" ), context );
+  const QString watershedBasinsOutputFile = parameterAsOutputLayer( parameters, QStringLiteral( "OUTPUT_WATERSHED_BASINS" ), context );
+
+  std::unique_ptr<QgsRasterFileWriter> filledDemWriter;
+  std::unique_ptr<QgsRasterDataProvider> filledDemDestProvider;
+
+  if ( !filledDemOutputFile.isEmpty() )
+  {
+    const QFileInfo fi( filledDemOutputFile );
+    const QString outputFormat = QgsRasterFileWriter::driverForExtension( fi.suffix() );
+
+    filledDemWriter = std::make_unique<QgsRasterFileWriter>( filledDemOutputFile );
+    filledDemWriter->setOutputProviderKey( QStringLiteral( "gdal" ) );
+    if ( !createOptions.isEmpty() )
+    {
+      filledDemWriter->setCreationOptions( createOptions.split( '|' ) );
+    }
+    filledDemWriter->setOutputFormat( outputFormat );
+
+    filledDemDestProvider.reset( filledDemWriter->createOneBandRaster( mDataType, mLayerWidth, mLayerHeight, mExtent, mCrs ) );
+
+    if ( !filledDemDestProvider )
+      throw QgsProcessingException( QObject::tr( "Could not create raster output: %1" ).arg( filledDemOutputFile ) );
+    if ( !filledDemDestProvider->isValid() )
+      throw QgsProcessingException( QObject::tr( "Could not create raster output %1: %2" ).arg( filledDemOutputFile, filledDemDestProvider->error().message( QgsErrorMessage::Text ) ) );
+
+    filledDemDestProvider->setNoDataValue( 1, mNoData );
+    filledDemDestProvider->setEditable( true );
+  }
+
+  std::unique_ptr<QgsRasterFileWriter> flowDirectionsWriter;
+  std::unique_ptr<QgsRasterDataProvider> flowDirectionsDestProvider;
+
+  if ( !flowDirectionsOutputFile.isEmpty() )
+  {
+    const QFileInfo fi( flowDirectionsOutputFile );
+    const QString outputFormat = QgsRasterFileWriter::driverForExtension( fi.suffix() );
+
+    flowDirectionsWriter = std::make_unique<QgsRasterFileWriter>( flowDirectionsOutputFile );
+    flowDirectionsWriter->setOutputProviderKey( QStringLiteral( "gdal" ) );
```

Comment:
Maybe we should apply creation options to this output too?

Reply:
Do we have capacity to set those per output? I was thinking that the options for the (likely float) dem output probably won't make sense when applied to the int types used for the other outputs...?

Reply:
We can create several parameters: one for each output. But probably this can be a bit confusing for users.

Reply:
Something like we already have for feature source inputs? Sounds like a good idea to me. Later on we can also add some other settings to it, e.g. nodata value.

Reply:
Sounds good!.. just another thing to add to the ever-growing wishlist :upside_down_face: